Add support for serving files directly from object storage via presigned URL redirects

stv0g opened the pull request:
This reduces the load in the Django app.

Tested on a Ceph RadosGW cluster running via rook.io in Kubernetes.

konstin commented:
You shouldn't use the django file serving backend in prod at all, but configure a redirect from nginx to minio as it is shown in `etc/nginx.conf`

konstin commented:
Unfortunately I have no clue about kubernetes, so I can't say anything about that. My understanding was that minio is a "kubernetes native" object storage, so it would be nice to integrate into a cluster (and possible to connect with disk/backblaze/s3)

konstin commented:
BTW have you seen https://github.com/meine-stadt-transparent/meine-stadt-transparent/tree/main/etc/dev? It's what I for development and this already uses minio through nginx instead of django/python; I've mostly kept the python version around for making it possible to develop or run tests without an nginx running

stv0g commented:
Thanks for the hints :)

stv0g commented:
I've think I will keep serving redirects (presigned URLs) via Django.
This has mainly two advantages for us:

1. I dont need allow anonymous access to the S3 bucket. The presigned URLs used in the redirect are only valid for a couple of minutes
2. I can include custom HTTP headers in the presigned URL. As a result S3 delivers the files with proper Content-type and Content-disposition header which allows me to serve them with their original filename when downloaded.

I also think that the overhead is neglegible as the actual file is still served directly from S3 rather then passing through Gunicorn.

My question would be: would you consider to merge this in MST or should we keep it in our fork for the long term?

konstin commented on mainapp/views/views.py:
```diff
@@ -243,18 +247,51 @@ def file_serve_proxy(
-    response = HttpResponse(minio_file.read())
+    name, ext = splitext(file.filename)
+    if name.isnumeric() and file.name:
+        filename = f'{file.name}_{name}{ext}'
```
is file.name save to use here, both regarding used characters and length?

stv0g replied:
Seems like this is not an easy topic.

There is [RFC5987](https://www.rfc-editor.org/rfc/rfc5987) which defines the encoding of UTF-8 strings in Header fields.

Also it seems like the filename needs to be percentage-encoded.

See also: https://stackoverflow.com/questions/93551/how-to-encode-the-filename-parameter-of-content-disposition-header-in-http

I will push a fix in a minute.

konstin commented on mainapp/views/views.py:
```diff
@@ -243,18 +247,51 @@ def file_serve_proxy(
+
+        mc = minio_client()
+        old_base_url = mc._base_url
+        mc._base_url = minio.api.BaseURL(("https://" if secure else "http://") + endpoint, settings.MINIO_REGION)
```
please avoid using private variables. in this case you can pass the settings on the `minio_client` function or just construct a new minio client object in the spot

stv0g replied:
Agreed. That was a nasty workaround.

I fixed it by creating two Minio instances of which one uses the public S3 endpoint of configured with MINIO_PUBLIC_HOST. This instance is only used for generating the presigned URLs. All the actual S3 operations are targeting always the default endpoint.

konstin reviewed (approved):
i didn't realize at first that this is actually using a redirect and that python needs to be involved to build the URL, so this is a nice addition!

stv0g commented:
Hi @konstin,

thanks for merging the other PRs :)

I have now rebased this PR on the latest state of `main`, addressed your comments and fixed the CI tests.

Cheers,
Steffen

konstin commented on mainapp/views/views.py:
```diff
@@ -242,18 +244,45 @@ def file_serve_proxy(
-    response["Content-Type"] = minio_file.headers["Content-Type"]
+    headers = {
+        # Encoding according to RFC5987
+        "Content-Disposition": f"attachment; filename=\"{quote(filename_safe)}\"; filename*=UTF-8''{quote(filename)}",
```
That's smarter than what I'd have used :+1:

konstin commented:
Thank you!
